In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\Usha Bharati\Downloads\Chocolate Bar Ratings'

In [3]:
df=pd.read_csv(os.path.join(path,'02 Data','Original Data','flavors_of_cacao.csv'),index_col=False)

In [4]:
df.head()

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                   A. Morin                       Agua Grande  1876   
1                   A. Morin                             Kpime  1676   
2                   A. Morin                            Atsane  1676   
3                   A. Morin                             Akata  1680   
4                   A. Morin                            Quilla  1704   

   Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0          2016            63%            France    3.75              
1          2015            70%            France    2.75              
2          2015            70%            France    3.00              
3          2015            70%            France    3.50              
4          2015            70%            France    3.50              

  Broad Bean\nOrigin  
0           Sao Tome  
1               Togo  
2               Togo  
3               Togo  
4               Peru

In [5]:
df.shape

(1795, 9)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 9 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Company 
(Maker-if known)         1795 non-null   object 
 1   Specific Bean Origin
or Bar Name  1795 non-null   object 
 2   REF                               1795 non-null   int64  
 3   Review
Date                       1795 non-null   int64  
 4   Cocoa
Percent                     1795 non-null   object 
 5   Company
Location                  1795 non-null   object 
 6   Rating                            1795 non-null   float64
 7   Bean
Type                         1794 non-null   object 
 8   Broad Bean
Origin                 1794 non-null   object 
dtypes: float64(1), int64(2), object(6)
memory usage: 126.3+ KB


In [7]:
df.dtypes

Company \n(Maker-if known)            object
Specific Bean Origin\nor Bar Name     object
REF                                    int64
Review\nDate                           int64
Cocoa\nPercent                        object
Company\nLocation                     object
Rating                               float64
Bean\nType                            object
Broad Bean\nOrigin                    object
dtype: object

In [8]:
df.describe()

REF  Review\nDate       Rating
count  1795.000000   1795.000000  1795.000000
mean   1035.904735   2012.325348     3.185933
std     552.886365      2.927210     0.478062
min       5.000000   2006.000000     1.000000
25%     576.000000   2010.000000     2.875000
50%    1069.000000   2013.000000     3.250000
75%    1502.000000   2015.000000     3.500000
max    1952.000000   2017.000000     5.000000

In [9]:
df.isnull().sum()

Company \n(Maker-if known)           0
Specific Bean Origin\nor Bar Name    0
REF                                  0
Review\nDate                         0
Cocoa\nPercent                       0
Company\nLocation                    0
Rating                               0
Bean\nType                           1
Broad Bean\nOrigin                   1
dtype: int64

In [10]:
df[df.duplicated()]

Empty DataFrame
Columns: [Company 
(Maker-if known), Specific Bean Origin
or Bar Name, REF, Review
Date, Cocoa
Percent, Company
Location, Rating, Bean
Type, Broad Bean
Origin]
Index: []

### With preliminary observation, it can be said that the dataset needs lot of cleaning.
### The names are very inconsistent and messy.
### There are two null values. One in each column of Bean\nType and Broad Bean\nOrigin.
### Out of 4 numerical columns- 'REF', 'Review\nDate','Cocoa\nPercent','Rating'- only 'Cocoa\nPercent' has data type as object. So, it needs to be changed to float datatype.
### There are no duplicate values in any column.
### Dataset shows only two null values.

## Renaming columns for consistency

In [11]:
# Renaming columns with appropriate names
df.rename(columns= {'Company\xa0\n(Maker-if known)':'Company',
                   'Specific Bean Origin\nor Bar Name':'Chocolate_origin',
                   'REF':'Review_number',
                   'Review\nDate':'Review_year',
                    'Cocoa\nPercent':'Cocoa_percent',
                    'Company\nLocation':'Company_location',
                    'Bean\nType':'Bean_type',
                     'Broad Bean\nOrigin':'Bean_origin'},inplace=True)

In [12]:
# Checking for corrected names of columns
df.head()

Company Chocolate_origin  Review_number  Review_year Cocoa_percent  \
0  A. Morin      Agua Grande           1876         2016           63%   
1  A. Morin            Kpime           1676         2015           70%   
2  A. Morin           Atsane           1676         2015           70%   
3  A. Morin            Akata           1680         2015           70%   
4  A. Morin           Quilla           1704         2015           70%   

  Company_location  Rating Bean_type Bean_origin  
0           France    3.75              Sao Tome  
1           France    2.75                  Togo  
2           France    3.00                  Togo  
3           France    3.50                  Togo  
4           France    3.50                  Peru

## Looking at individual columns for inconsistencies. 

### 1. 'Company' column

In [13]:
df['Company'].unique()

array(['A. Morin', 'Acalli', 'Adi', 'Aequare (Gianduja)', 'Ah Cacao',
       "Akesson's (Pralus)", 'Alain Ducasse', 'Alexandre',
       'Altus aka Cao Artisan', 'Amano', 'Amatller (Simon Coll)',
       'Amazona', 'Ambrosia', 'Amedei', 'AMMA', 'Anahata', 'Animas',
       'Ara', 'Arete', 'Artisan du Chocolat',
       'Artisan du Chocolat (Casa Luker)', 'Askinosie', 'Bahen & Co.',
       'Bakau', 'Bar Au Chocolat', "Baravelli's", 'Batch', 'Beau Cacao',
       'Beehive', 'Belcolade', 'Bellflower', 'Belyzium', 'Benoit Nihant',
       'Bernachon', 'Beschle (Felchlin)', 'Bisou', 'Bittersweet Origins',
       'Black Mountain', 'Black River (A. Morin)', 'Blanxart',
       'Blue Bandana', 'Bonnat', 'Bouga Cacao (Tulicorp)', 'Bowler Man',
       "Brasstown aka It's Chocolate", 'Brazen', 'Breeze Mill', 'Bright',
       'Britarev', 'Bronx Grrl Chocolate', 'Burnt Fork Bend',
       'Cacao Arabuco', 'Cacao Atlanta', 'Cacao Barry', 'Cacao de Origen',
       'Cacao de Origin', 'Cacao Hunters', 'Cacao M

In [14]:
# Changing one company name containing character
df['Company']=df['Company'].replace('Na�ve','Naive')

### 2.'Chocolate_origin' column

In [15]:
df['Chocolate_origin'].unique()

array(['Agua Grande', 'Kpime', 'Atsane', ..., 'Indianer, Raw',
       'Kerala State', 'Brazil, Mitzi Blue'], dtype=object)

#### This column does not need much changes as such.

### 3. 'Cocoa_percent' column

In [16]:
# Changing the data type of Cocoa_percent from object, by removing % symbol and converting to numerical-float.

df['Cocoa_percent'] = df['Cocoa_percent'].str.rstrip('%').astype(float)

In [17]:
# Checking for correction
df.head()

Company Chocolate_origin  Review_number  Review_year  Cocoa_percent  \
0  A. Morin      Agua Grande           1876         2016           63.0   
1  A. Morin            Kpime           1676         2015           70.0   
2  A. Morin           Atsane           1676         2015           70.0   
3  A. Morin            Akata           1680         2015           70.0   
4  A. Morin           Quilla           1704         2015           70.0   

  Company_location  Rating Bean_type Bean_origin  
0           France    3.75              Sao Tome  
1           France    2.75                  Togo  
2           France    3.00                  Togo  
3           France    3.50                  Togo  
4           France    3.50                  Peru

### 4. 'Company_location' column

In [18]:
# Looking for unique values
df['Company_location'].unique()

array(['France', 'U.S.A.', 'Fiji', 'Ecuador', 'Mexico', 'Switzerland',
       'Netherlands', 'Spain', 'Peru', 'Canada', 'Italy', 'Brazil',
       'U.K.', 'Australia', 'Wales', 'Belgium', 'Germany', 'Russia',
       'Puerto Rico', 'Venezuela', 'Colombia', 'Japan', 'New Zealand',
       'Costa Rica', 'South Korea', 'Amsterdam', 'Scotland', 'Martinique',
       'Sao Tome', 'Argentina', 'Guatemala', 'South Africa', 'Bolivia',
       'St. Lucia', 'Portugal', 'Singapore', 'Denmark', 'Vietnam',
       'Grenada', 'Israel', 'India', 'Czech Republic',
       'Domincan Republic', 'Finland', 'Madagascar', 'Philippines',
       'Sweden', 'Poland', 'Austria', 'Honduras', 'Nicaragua',
       'Lithuania', 'Niacragua', 'Chile', 'Ghana', 'Iceland', 'Eucador',
       'Hungary', 'Suriname', 'Ireland'], dtype=object)

#### There are some country names misspelled, so need to be corrected.


In [19]:
# Replace incorrect country names
corrections = {
    'U.S.A.': 'USA',
    'Eucador': 'Ecuador',
    'Niacragua': 'Nicaragua',
    'Domincan Republic': 'Dominican Republic'
}

df['Company_location'] = df['Company_location'].replace(corrections)

In [20]:
# Checking for changed corrected names
df['Company_location'].unique()

array(['France', 'USA', 'Fiji', 'Ecuador', 'Mexico', 'Switzerland',
       'Netherlands', 'Spain', 'Peru', 'Canada', 'Italy', 'Brazil',
       'U.K.', 'Australia', 'Wales', 'Belgium', 'Germany', 'Russia',
       'Puerto Rico', 'Venezuela', 'Colombia', 'Japan', 'New Zealand',
       'Costa Rica', 'South Korea', 'Amsterdam', 'Scotland', 'Martinique',
       'Sao Tome', 'Argentina', 'Guatemala', 'South Africa', 'Bolivia',
       'St. Lucia', 'Portugal', 'Singapore', 'Denmark', 'Vietnam',
       'Grenada', 'Israel', 'India', 'Czech Republic',
       'Dominican Republic', 'Finland', 'Madagascar', 'Philippines',
       'Sweden', 'Poland', 'Austria', 'Honduras', 'Nicaragua',
       'Lithuania', 'Chile', 'Ghana', 'Iceland', 'Hungary', 'Suriname',
       'Ireland'], dtype=object)

### 5.'Rating' column

In [21]:
df['Rating'].value_counts()

Rating
3.50    392
3.00    341
3.25    303
2.75    259
3.75    210
2.50    127
4.00     98
2.00     32
2.25     14
1.50     10
1.00      4
1.75      3
5.00      2
Name: count, dtype: int64

#### There are rating values from 1.00 to 5.00 with continuous values in between. No changes are made in this column as they seem to be required in as is condition.

### 6.'Bean_type' column

In [22]:
# Looking for unique values in this column
df['Bean_type'].unique()

array(['\xa0', 'Criollo', 'Trinitario', 'Forastero (Arriba)', 'Forastero',
       'Forastero (Nacional)', 'Criollo, Trinitario',
       'Criollo (Porcelana)', 'Blend', 'Trinitario (85% Criollo)',
       'Forastero (Catongo)', 'Forastero (Parazinho)',
       'Trinitario, Criollo', 'CCN51', 'Criollo (Ocumare)', 'Nacional',
       'Criollo (Ocumare 61)', 'Criollo (Ocumare 77)',
       'Criollo (Ocumare 67)', 'Criollo (Wild)', 'Beniano', 'Amazon mix',
       'Trinitario, Forastero', 'Forastero (Arriba) ASS', 'Criollo, +',
       'Amazon', 'Amazon, ICS', 'EET', 'Blend-Forastero,Criollo',
       'Trinitario (Scavina)', 'Criollo, Forastero', 'Matina',
       'Forastero(Arriba, CCN)', 'Nacional (Arriba)',
       'Forastero (Arriba) ASSS', 'Forastero, Trinitario',
       'Forastero (Amelonado)', nan, 'Trinitario, Nacional',
       'Trinitario (Amelonado)', 'Trinitario, TCGA', 'Criollo (Amarru)'],
      dtype=object)

In [23]:
df['Bean_type'].value_counts()

Bean_type
                            887
Trinitario                  419
Criollo                     153
Forastero                    87
Forastero (Nacional)         52
Blend                        41
Criollo, Trinitario          39
Forastero (Arriba)           37
Criollo (Porcelana)          10
Trinitario, Criollo           9
Forastero (Parazinho)         8
Forastero (Arriba) ASS        6
Nacional (Arriba)             3
Matina                        3
EET                           3
Beniano                       3
Criollo (Ocumare 61)          2
Trinitario, Forastero         2
Trinitario (85% Criollo)      2
Forastero (Catongo)           2
Criollo, Forastero            2
Amazon, ICS                   2
Criollo (Amarru)              2
Amazon mix                    2
Nacional                      2
Criollo (Ocumare)             1
Criollo (Ocumare 67)          1
Trinitario, TCGA              1
Trinitario (Amelonado)        1
Trinitario, Nacional          1
Forastero (Amelonado)         

#### There appears to be- '\xa0' with no entry for almost 887 cells. These appear to be missing or unknown values. Replacing '\xa0' with 'Unknown' would make the data more readable.

In [24]:
# Replacing '\xa0' witg 'Unknown'
df['Bean_type']=df['Bean_type'].replace('\xa0','Unknown')

In [25]:
df['Bean_type'].value_counts()

Bean_type
Unknown                     887
Trinitario                  419
Criollo                     153
Forastero                    87
Forastero (Nacional)         52
Blend                        41
Criollo, Trinitario          39
Forastero (Arriba)           37
Criollo (Porcelana)          10
Trinitario, Criollo           9
Forastero (Parazinho)         8
Forastero (Arriba) ASS        6
Nacional (Arriba)             3
Matina                        3
EET                           3
Beniano                       3
Criollo (Ocumare 61)          2
Trinitario, Forastero         2
Trinitario (85% Criollo)      2
Forastero (Catongo)           2
Criollo, Forastero            2
Amazon, ICS                   2
Criollo (Amarru)              2
Amazon mix                    2
Nacional                      2
Criollo (Ocumare)             1
Criollo (Ocumare 67)          1
Trinitario, TCGA              1
Trinitario (Amelonado)        1
Trinitario, Nacional          1
Forastero (Amelonado)         

### 7. 'Bean_origin' column

In [26]:
df['Bean_origin'].unique()

array(['Sao Tome', 'Togo', 'Peru', 'Venezuela', 'Cuba', 'Panama',
       'Madagascar', 'Brazil', 'Ecuador', 'Colombia', 'Burma',
       'Papua New Guinea', 'Bolivia', 'Fiji', 'Mexico', 'Indonesia',
       'Trinidad', 'Vietnam', 'Nicaragua', 'Tanzania',
       'Dominican Republic', 'Ghana', 'Belize', '\xa0', 'Jamaica',
       'Grenada', 'Guatemala', 'Honduras', 'Costa Rica',
       'Domincan Republic', 'Haiti', 'Congo', 'Philippines', 'Malaysia',
       'Dominican Rep., Bali', 'Venez,Africa,Brasil,Peru,Mex', 'Gabon',
       'Ivory Coast', 'Carribean', 'Sri Lanka', 'Puerto Rico', 'Uganda',
       'Martinique', 'Sao Tome & Principe', 'Vanuatu', 'Australia',
       'Liberia', 'Ecuador, Costa Rica', 'West Africa', 'Hawaii',
       'St. Lucia', 'Cost Rica, Ven', 'Peru, Madagascar',
       'Venezuela, Trinidad', 'Trinidad, Tobago',
       'Ven, Trinidad, Ecuador', 'South America, Africa', 'India',
       'Africa, Carribean, C. Am.', 'Tobago', 'Ven., Indonesia, Ecuad.',
       'Trinidad-Tobago

####  This column also has -'\xa0' some missing values, so they are replaced with 'Unknown'.
#### There are misspelt names and with character inbetween, which need correction.

In [27]:
correction = {
    '\xa0': 'Unknown',
    'Venez,Africa,Brasil,Peru,Mex': 'Venezuela,Africa,Brazil,Peru,Mexico',
    'Dominican Rep.': 'Dominican Republic',
    'Domincan Republic': 'Dominican Republic',
    'Cost Rica, Ven': 'Costa Rica, Venezuela',
    'Ven, Trinidad, Ecuador': 'Venezuela, Trinidad, Ecuador',
    'Sao Tome & Principe': 'Sao Tome and Principe',
    'Trinidad-Tobago': 'Trinidad and Tobago',
    'Carribean': 'Caribbean',
    'Dominican Rep., Bali': 'Dominican Republic, Bali',
    'Venezuela/ Ghana': 'Venezuela, Ghana',
    'Sri Lanka': 'Sri Lanka',
    'Ven., Trinidad, Mad.': 'Venezuela, Trinidad, Madagascar',
    'Ven, Bolivia, D.R.': 'Venezuela, Bolivia, Dominican Republic',
    'Ghana & Madagascar': 'Ghana, Madagascar',
    'Venezuela, Carribean': 'Venezuela, Caribbean',
    'Guat., D.R., Peru, Mad., PNG': 'Guatemala, Dominican Republic, Peru, Madagascar, Papua New Guinea',
    'Madagascar & Ecuador': 'Madagascar, Ecuador',
    'Ven.,Ecu.,Peru,Nic.': 'Venezuela, Ecuador, Peru, Nicaragua',
    'Ven., Trinidad, Mad.': 'Venezuela, Trinidad, Madagascar',
    'Gre., PNG, Haw., Haiti, Mad': 'Grenada, Papua New Guinea, Hawaii, Haiti, Madagascar',
    'Peru, Dom. Rep': 'Peru, Dominican Republic',
    'Venezuela, Trinidad': 'Venezuela, Trinidad',
    'Peru, Ecuador': 'Peru, Ecuador',
    'Peru, Madagascar': 'Peru, Madagascar',
    'Venezuela, Ghana': 'Venezuela, Ghana',
    'Trinidad, Tobago': 'Trinidad, Tobago',
    'Peru(SMartin,Pangoa,nacional)': 'Peru (SMartin, Pangoa, nacional)',
    'Principe': 'Principe',
    'Central and S. America': 'Central and South America',
    'Venez,Africa,Brasil,Peru,Mex': 'Venezuela, Africa, Brazil, Peru, Mexico',
    'PNG, Vanuatu, Mad': 'Papua New Guinea, Vanuatu, Madagascar'
}
df['Bean_origin'] = df['Bean_origin'].replace(corrections)

In [28]:
df['Bean_origin'].value_counts()

Bean_origin
Venezuela                214
Ecuador                  193
Dominican Republic       166
Peru                     165
Madagascar               145
                        ... 
Peru, Belize               1
Peru, Mad., Dom. Rep.      1
PNG, Vanuatu, Mad          1
Trinidad, Ecuador          1
Venezuela, Carribean       1
Name: count, Length: 99, dtype: int64

#### Descrptive statistics of numerical columns

In [29]:
df.describe()

Review_number  Review_year  Cocoa_percent       Rating
count    1795.000000  1795.000000    1795.000000  1795.000000
mean     1035.904735  2012.325348      71.698329     3.185933
std       552.886365     2.927210       6.323118     0.478062
min         5.000000  2006.000000      42.000000     1.000000
25%       576.000000  2010.000000      70.000000     2.875000
50%      1069.000000  2013.000000      70.000000     3.250000
75%      1502.000000  2015.000000      75.000000     3.500000
max      1952.000000  2017.000000     100.000000     5.000000

#### Review of chocolate bars have been conducted from 2006 to 2017.
#### Highest and lowest cocoa percentages in chocolate bars are 100% and 42% respectively, With average containing 71.6% cocoa.
#### Highest and lowest ratings being 5 and 1 respectively and average rating being 3.1

## Export Data

In [31]:
# Export the cleaned data 
df.to_csv(os.path.join(path, '02 Data','Prepared Data','df_cleaned.csv'), index=False)